crates/wasi/src/ctx.rs

```diff
@@ -156,37 +232,105 @@ impl WasiCtxBuilder {
         self.envs(&std::env::vars().collect::<Vec<(String, String)>>())
     }
 
+    /// Appends a list of arguments to the argument array to pass to wasm.
     pub fn args(&mut self, args: &[impl AsRef<str>]) -> &mut Self {
         self.args.extend(args.iter().map(|a| a.as_ref().to_owned()));
         self
     }
 
+    /// Appends a single argument to get passed to wasm.
     pub fn arg(&mut self, arg: impl AsRef<str>) -> &mut Self {
         self.args.push(arg.as_ref().to_owned());
         self
     }
 
+    /// Appends all host process arguments to the list of arguments to get
+    /// passed to wasm.
     pub fn inherit_args(&mut self) -> &mut Self {
         self.args(&std::env::args().collect::<Vec<String>>())
     }
 
+    /// Configures a "preopened directory" to be available to WebAssembly.
+    ///
+    /// By default WebAssembly does not have access to the filesystem because
+    /// the are no preopened directories. All filesystem operations, such as
+    /// opening a file, are done through a preexisting handle. This means that
+    /// to provide WebAssembly access to a directory it must be configured
+    /// through this API.
+    ///
+    /// WASI will also prevent access outside of files provided here. For
+    /// example `..` can't be used to traverse up from the `dir` provided here
+    /// to the containing directory.
+    ///
+    /// * `host_path` - a path to a directory on the host to open and make
+    ///   accessible to WebAssembly. Note that the name of this directory in the
+    ///   guest is configured with `guest_path` below.
+    /// * `perms` - this is the permissions that wasm will have to operate on
+    ///   `dir`. This can be used, for example, to provide readonly access to a
+    ///   directory.
+    /// * `file_perms` - similar to `perms` but corresponds to the maximum set
+    ///   of permissions that can be used for any file in this directory.
+    /// * `guest_path` - the name of the preopened directory from WebAssembly's
+    ///   perspective. Note that this does not need to match the host's name for
+    ///   the directory.
+    ///
+    /// # Errors
+    ///
+    /// This method will return an error if `host_path` cannot be opened.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use wasmtime_wasi::{WasiCtxBuilder, DirPerms, FilePerms};
+    ///
+    /// # fn main() {}
+    /// # fn foo() -> wasmtime::Result<()> {
+    /// let mut wasi = WasiCtxBuilder::new();
+    ///
+    /// // Make `./host-directory` available in the guest as `.`
+    /// wasi.preopened_dir("./host-directory", DirPerms::all(), FilePerms::all(), ".");
+    ///
+    /// // Make `./readonly` available in the guest as `./ro`
+    /// wasi.preopened_dir("./readonly", DirPerms::READ, FilePerms::READ, "./ro");
+    /// # Ok(())
+    /// # }
+    /// ```
     pub fn preopened_dir(
         &mut self,
-        dir: cap_std::fs::Dir,
+        host_path: impl AsRef<Path>,
         perms: DirPerms,
```

Comment:
This always bothered me that this is called `perms` instead of `dir_perms` to contrast with `file_perms`. What do we think about renaming that arg?